csrc/optimization/consecutive_cast_pass.cpp

```diff
@@ -0,0 +1,99 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <optimization/consecutive_cast_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+void castOptimizationPass(Fusion* fusion) {
+  auto is_cast_op = [](Expr* expr) {
+    if (expr->isA<UnaryOp>()) {
+      auto op = expr->as<UnaryOp>();
+      if (op->getUnaryOpType() == UnaryOpType::Cast) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  // NOTE: not the most efficient pass
+  std::unordered_map<Val*, Val*> replacement_map;
+  for (auto expr : fusion->exprs()) {
+    if (is_cast_op(expr)) {
+      bool mutated = false;
+      while (true) {
+        // in the loop, we just repetitively skip consecutive casts.
+        auto intermediate_cast = expr->input(0);
+        auto prev_expr = intermediate_cast->definition();
+        if (prev_expr != nullptr && is_cast_op(prev_expr)) {
+          auto original_dtype = prev_expr->input(0)->getDataType().value();
+          auto intermediate_dtype = intermediate_cast->getDataType().value();
+          auto out_dtype = expr->output(0)->getDataType().value();
+          // cases where skipping the intermediate cast is relatively safe,
```

Comment:
> What does "relatively safe" exactly mean? Are these rules an accepted convention in PyTorch?

Well, it's a vague term and that's because the purpose we are trying to serve here: https://github.com/NVIDIA/Fuser/issues/335#issuecomment-1548097006

Xiang also brought this up and arguably we should have limited the cast elimination to only when it's strictly a no-op case when the removal doesn't change the output at all, but that won't help us with the perf issue that @liqiangxl is fighting.

> I also don't understand what "relaxed" means.

So you are referring to this one vvv
```
          //   2. intermediate_dtype is more relaxed than out_dtype. e.g. a
          //   floating point vs. integral;
         ...
               (isFloatingPointType(intermediate_dtype) &&
                isIntegralType(out_dtype)))) {
```
err. I don't have a good example to defend it. I'll remove that case.

Reply:
Note that the true root cause of this performance issue is #343. The cast should be a cheap operation and should not affect the overall performance. Removing the cast would avoid #343 and make the performance of #335 look much better, but we still need to fix #343.

Reply:
Does this mean that we don't want to have the WAR in this pass? I'm perfectly fine with refactor the pass to only fold casts that's a strict no-op, which feels like a more general pass that can be useful in more places. And also a lot more easier to justify using it.

Reply:
Hmm, why does the original case have the series of cast? Is it because how it's written by the user? Or are they automatically inserted?

Reply:
It's a combination of the two.

A chunk of our ops don't do fp16 math, so we have a logic that promotes inputs to fp32 and demote it back to fp16 afterwards. So if user tries to force computation in fp16 precision, we'll ended up with a cast to fp16 by user and then a cast back to fp32 by our type promotion logic.

Reply:
Sounds like we are trying to remove cast ops that we prematurely inserted. I wonder if it would make more sense to "fix" the initial insertion algorithm rather than fixing it up later.

Reply:
> I wonder if it would make more sense to "fix" the initial insertion algorithm rather than fixing it up later.

`Ensuring customer graph being in good shape` is a relatively high maintenance approach in my opinion.

Having said that, I'm hearing this as that we want to stay safe on the optimization pass and steer clear from graph mutations that could affect the outputs. I think it's a fair ask and I'm OK moving forward in that direction, given that we'll eventually push through #343 and cast optimization should no longer been needed desperately for performance.

Reply:
Sorry, maybe I misinterpreted your previous comment. I thought this:

> so we have a logic that promotes inputs to fp32 and demote it back to fp16 afterwards.

is done by us when defining a fusion, specifically `promoteValues` in type_promotion.cpp. If so, could we change the promotion algorithm to not insert back-and-forth cast ops?

Reply:
> Note that the true root cause of this performance issue is #343. The cast should be a cheap operation and should not affect the overall performance. Removing the cast would avoid #343 and make the performance of #335 look much better, but we still need to fix #343.

So this means we still want to implement approach-2 mensioned in #343 ? @naoyam

Reply:
Yes. Didn't you also find a different persistent buffer issue when you worked on the layer norm backward?

Reply:
> Yes. Didn't you also find a different persistent buffer issue when you worked on the layer norm backward?

I don't remember that. The issue back there was how to estimate the size of the persistet buffers and it was already solved.

Reply:
Wasn't there something that wasn't specific to the layer norm backward but should be considered for more general cases?

Reply:
I think this one: https://github.com/csarofeen/pytorch/pull/2400#issuecomment-1439310912

Reply:
got you. we do have an issue tracking it in the old repo. https://github.com/csarofeen/pytorch/issues/2525

Reply:
Can you please move the issue to this repo?

Reply:
moved to https://github.com/NVIDIA/Fuser/issues/402